### ISSUE TITLE
Reject requests with read point older than the history cutoff timestamp

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#288]
A bug could happen in transactional reads when the hybrid time is assigned at the YQL node and not on the tablet itself. To fix this:
- Retain a longer history by default (a few seconds).
- Reject requests that have the hybrid time older than the current history cutoff that could have already been garbage collected.

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-288-analysis.md`):

# Analysis: yugabyte/yugabyte-db#288 (no Jira cross-link)

Issue: "Reject requests with read point older than the history cutoff timestamp"
(kind/bug, filed 2018-05-15, OPEN, zero comments). Backlog-triage validation of a
DocDB correctness concern: a transactional read whose hybrid time was assigned at
the YQL node could arrive at a tablet after a compaction had already
garbage-collected history past that read point, silently returning incorrect
data. The issue asks to (a) retain longer history by default and (b) reject
reads below the history cutoff.

Reviewed on master `b89bc62335` (2026-08-14), read-only checkout, 2026-08-16.

## Verdict

FIXED (app verdict: Fixed) — the exact rejection requested was implemented by
commit `23f60107cb1845f4d8232207b2820fd616b47a43` (2019-01-09, "#704 ENG-4422
Return a \"Snapshot too old\" error on attempts to read compacted-away data,
e.g. during very long-running transactions", Phorge D5912) and is enforced and
regression-tested on current master. #288 predates and duplicates #704, which
was closed 2019-01-22; #288 was simply never closed. In skill vocabulary this is
"not a bug on master / already fixed" — no reproduction of a defect is possible
because the defect no longer exists.

## Root cause (current-state mechanism)

- Enforcement: `src/yb/tablet/tablet_retention_policy.cc:132-164` —
  `TabletRetentionPolicy::RegisterReaderTimestamp(HybridTime)` compares the read
  timestamp against `GetEarliestAllowedReadHt()` (the committed history cutoff,
  `tablet_retention_policy.cc:177-192`) and returns
  `STATUS(SnapshotTooOld, "Snapshot too old. Read point: $0, earliest read time
  allowed: $1 ...", TransactionError(kSnapshotTooOld))` at
  `tablet_retention_policy.cc:145-156`. Status code `SnapshotTooOld` (code 30)
  is defined at `src/yb/util/status_codes.h:47`.
- Wiring: every read acquires a `ScopedReadOperation`;
  `ScopedReadOperation::Create` calls `RegisterReaderTimestamp` at
  `src/yb/tablet/tablet.cc:6344` (function at `tablet.cc:6335-6347`). Callers:
  tserver read path `src/yb/tserver/read_query.cc:750` (backfill reads) and
  `:771-772` (normal reads, including client/YQL-node-supplied `read_time_` —
  precisely the scenario in the issue body); write-side read times
  `src/yb/tablet/write_query.cc:1102`.
- Reader pinning (prevents the race in the other direction):
  registered readers are inserted into `active_readers_`
  (`tablet_retention_policy.cc:162`) and `SanitizeHistoryCutoff`
  (`tablet_retention_policy.cc:219-247`, reader clamp at `:228-233`) caps any
  proposed compaction history cutoff at the oldest active read, so an in-flight
  read can never be invalidated by a concurrent compaction.
- Retention default (issue ask "a few seconds"):
  `timestamp_history_retention_interval_sec` defaults to **900 s** at
  `src/yb/common/common_flags.cc:361` (the 2019 fix shipped 10 s); syscatalog
  retention is 4 h (`tablet_retention_policy.cc:47`).
- Durability: the committed cutoff is persisted in RocksDB metadata via the
  consensus "frontiers" mechanism and restored on restart (per `23f60107cb`
  summary; `src/yb/docdb/consensus_frontier.{h,cc}` in its diffstat), so the
  guard survives tserver restarts.

Master affected: **no** — the guard is present, on by default, and tested.

## Landing commit & follow-ups

- `23f60107cb` — original fix (#704 / ENG-4422 / D5912), Mikhail Bautin,
  2019-01-09. Introduced `earliest_read_time_allowed` + SnapshotTooOld
  rejection + cutoff persistence.
- `d56ba9c061` — [#1032] history cutoff propagation from leader.
- `f1cc6b55d3` — [#32560] backfill table reads below history cutoff fail loud
  (`read_query.cc:742-760`, gated by `backfill_index_check_snapshot_too_old`).
- `9b4eb7c579` — [#30329] deliberate exemption: index tablets with
  `retain_delete_markers=true` (backfill in progress) skip the check — safe
  because compaction retains the data such readers need; documented at
  `tablet_retention_policy.cc:136-143`.

## Trigger matrix (predicted from source vs observed in source/tests)

| condition | predicted | observed |
|---|---|---|
| Read RPC, read_ht < committed history cutoff, normal table | rejected SnapshotTooOld | `read_query.cc:771-772` -> `tablet.cc:6344` -> `tablet_retention_policy.cc:135-156` returns SnapshotTooOld; asserted by `pg_packed_row-test.cc:294,680`, `pg_catalog_perf-test.cc:694-705` |
| Read RPC, read_ht >= cutoff | allowed; pins cutoff while active | `tablet_retention_policy.cc:162` inserts into `active_readers_`; `SanitizeHistoryCutoff:228-233` clamps cutoff to oldest reader |
| Backfill read of indexed table below cutoff | rejected (fail loud, no silent wrong index) | `read_query.cc:742-760`; `cassandra_cpp_driver-test.cc:2577`, `pg_index_backfill-test.cc:1366,1518` |
| Index tablet during its own backfill (retain_delete_markers) | exempt by design | `tablet_retention_policy.cc:144-160` (VLOG + allow), rationale comment `:136-143` |
| Write op needing read time below cutoff | rejected via same path | `write_query.cc:1102` -> `ScopedReadOperation::Create` |
| Default retention window | >= "a few seconds" asked | 900 s, `common_flags.cc:361` |

All cells observed as predicted; no divergence.

## Reproduction

N/A — no defect to reproduce on master; this is a code-inspection + git-archaeology
validation of a 2018 design-gap report with no failure signature. Rerunnable
evidence commands (from repo root, master `b89bc62335`):

```sh
gh issue view 288 --repo yugabyte/yugabyte-db --json title,body,state,createdAt  # OPEN, 2018-05-15
gh issue view 704 --repo yugabyte/yugabyte-db --json state,closedAt              # CLOSED 2019-01-22
rg -n 'Snapshot too old|SnapshotTooOld' src/yb/tablet/tablet_retention_policy.cc src/yb/util/status_codes.h
rg -n 'RegisterReaderTimestamp' src/yb/tablet/tablet.cc src/yb/tserver/read_query.cc
rg -n 'DEFINE_.*timestamp_history_retention_interval_sec' src/yb/common/common_flags.cc  # 900
git log --oneline --reverse -S'SnapshotTooOld' -- src/yb/tablet/tablet_retention_policy.cc src/yb/tablet/tablet.cc
# -> 23f60107cb "#704 ENG-4422 Return a 'Snapshot too old' error ..." (2019-01-09)
git show 23f60107cb --stat --format='%an %ad %s'
```

Runtime confirmation exists as in-tree regression tests (not run here; no local
builds per dispatch): `pg_packed_row-test.cc:294,680`,
`pg_catalog_perf-test.cc:694`, `pg_index_backfill-test.cc:1366`. The test flag
`TEST_committed_history_cutoff_initial_value_usec`
(`tablet_retention_policy.cc:59`) force-triggers the rejection deterministically.

## Versions tested

- master `b89bc6233565f83487971bcbef4bc5b8532a42f3` (2026-08-14) — source
  inspection; guard present. Master affected: no.
- Fix present since `23f60107cb` (2019-01-09), i.e. every release line from
  v1.1.x onward. No container matrix run — nothing version-specific to bisect.

## Proposed fix direction

No code change. Issue hygiene only: close #288 as fixed by
`23f60107cb`/D5912 (duplicate of #704, closed 2019-01-22). If maintainers want a
paper trail, a closing comment citing `tablet_retention_policy.cc:132-164` and
the commit suffices.

## Test strategy

Covered by existing tests exercising public entry points:
`pg_packed_row-test.cc` ("Snapshot too old" on read after aggressive
compaction), `pg_catalog_perf-test.cc` (response-cache renewal on SnapshotTooOld),
`pg_index_backfill-test.cc` / `cassandra_cpp_driver-test.cc` (backfill fail-loud).
No new tests needed.

## Confidentiality check

Clean — contains only public-repo facts: public issue numbers (#288, #704,
#1032, #30329, #32560), upstream commit hashes, in-tree file:line citations,
and public flag names. No customer data, PII, internal hostnames, or secrets.
